### PR TITLE
Switch GSL to MS GSL 4.0.0

### PR DESCRIPTION
**Pull request by edgchen1**
### Description

Switch from onnxruntime copy of gsl-lite to MS GSL 4.0.0.

### Motivation and Context

Update GSL to more recent version.
There are also some requests for onnxruntime to use MS GSL.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
Should we not add GSL via Fetch mechanims which allows optional fetching in case some subproject has already done that?
Because of XNNPAK CPUInfo we choose to disable it all together instead of reusing each others git repo.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -129,7 +129,6 @@ extern "C" {
 
 // Used in *.cc files. Almost as same as ORT_API_STATUS, except without ORT_MUST_USE_RESULT and ORT_EXPORT
 #define ORT_API_STATUS_IMPL(NAME, ...) \
-  GSL_SUPPRESS(r .11)                  \
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:132]: maybe we should keep this, but if so we shouldn't have a GSL dependency in the C API header

**Comment by edgchen1** [comment]
> Should we not add GSL via Fetch mechanims which allows optional fetching in case some subproject has already done that? Because of XNNPAK CPUInfo we choose to disable it all together instead of reusing each others git repo.

will look into this

**Comment by snnn** [comment]
> Should we not add GSL via Fetch mechanims which allows optional fetching in case some subproject has already done that? Because of XNNPAK CPUInfo we choose to disable it all together instead of reusing each others git repo.

I can address it in another PR. I'm working on a change that will convert most submodules to cmake Fetch. But, @edgchen1 , if you want to do it now, it would be fine too.

**Comment by snnn** [comment]
All pipeline passed!

File: cgmanifests/generated/cgmanifest.json [hunk 2]
```diff
@@ -163,7 +173,7 @@
       "component": {
         "type": "git",
         "git": {
-          "commitHash": "c220895fd1163c01f0a8b44229fb9e4fe0ae0958",
+          "commitHash": "fc645b7626ebf86530dbd82fbece74d457e7ae07",
```
Comment (edgchen1) [inline cgmanifests/generated/cgmanifest.json:176]: revert this change

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (snnn) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:132]: This macro is for internal uses only when implementing C API functions. So I'm less concerned.

File: onnxruntime/contrib_ops/cpu/attnlstm/bahdanau_attention.cc [hunk 3]
```diff
@@ -63,11 +63,11 @@ template <typename T>
 void BahdanauAttention<T>::PrepareMemory(
     const gsl::span<const T>& memory,
     const gsl::span<const int>& memory_sequence_lengths) {
-  std::copy(memory.cbegin(), memory.cend(), values_.begin());
```
Comment (snnn) [inline onnxruntime/contrib_ops/cpu/attnlstm/bahdanau_attention.cc:66]: This is why I think the migration to MS GSL the sooner the better. In the future, we may consider to create a safer wrapper of std::copy (and memcpy) with range check. Here when we convert spans to iterators, I believe the range information is lost, and the main benefit of using span is gone.

**Comment by edgchen1** [comment]
> > Should we not add GSL via Fetch mechanims which allows optional fetching in case some subproject has already done that? Because of XNNPAK CPUInfo we choose to disable it all together instead of reusing each others git repo.
> 
> I can address it in another PR. I'm working on a change that will convert most submodules to cmake Fetch. But, @edgchen1 , if you want to do it now, it would be fine too.

Thanks - I will try to make the change here.

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -1016,27 +1016,18 @@ if (CPUINFO_SUPPORTED)
...
-  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DGSL_TERMINATE_ON_CONTRACT_VIOLATION")
-else()
-  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DGSL_THROW_ON_CONTRACT_VIOLATION")
-endif()
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -DGSL_UNENFORCED_ON_CONTRACT_VIOLATION")
-set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -DGSL_UNENFORCED_ON_CONTRACT_VIOLATION")
-set(CMAKE_CXX_FLAGS_MINSIZEREL "${CMAKE_CXX_FLAGS_MINSIZEREL} -DGSL_UNENFORCED_ON_CONTRACT_VIOLATION")
+# GSL
+set(GSL_DIR "${REPO_ROOT}/cmake/external/GSL")
+add_subdirectory(${GSL_DIR} EXCLUDE_FROM_ALL)
```
Comment (snnn) [inline cmake/CMakeLists.txt:1021]: GSL is header only.  So we don't need to use their cmake file. Their cmake files may have things that we don't want. For example, there is a GitHub issue in their repo saying that their cmake file would propagate "-std=c++14" to the targets that link to GSL. But we want C++17 instead. We don't need their cmake target to help us setup any compile flag. Therefore there is no need to use their cmake file.

File: cmake/external/gsl.cmake [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+include(FetchContent)
+
+FetchContent_Declare(
+    GSL
+    GIT_REPOSITORY https://github.com/microsoft/gsl
```
Comment (snnn) [inline cmake/external/gsl.cmake:8]: You may use a https url so that we don't need to download all versions of gsl.

Comment (edgchen1) [reply cmake/external/gsl.cmake:8]: would GIT_SHALLOW help with that?

Comment (snnn) [reply cmake/external/gsl.cmake:8]: Then why not just use the URL https://github.com/microsoft/GSL/archive/refs/tags/v4.0.0.zip ? Like what we have in https://github.com/microsoft/onnxruntime/blob/main/cmake/external/abseil-cpp.cmake

Comment (edgchen1) [reply cmake/external/gsl.cmake:8]: It's based on what was recommended here:
https://github.com/microsoft/GSL/blob/1e0d0446dddfa019dae17ffff4398bf3b96ad6d4/README.md?plain=1#L205-L211

Also, I hope we can come up with a way to update cgmanifest entries automatically for these dependencies. This seems easier to parse than a URL.

Comment (snnn) [reply cmake/external/gsl.cmake:8]: Ok, We can discuss it later.

**Comment by snnn** [comment]
The four failures in Windows CPU CI pipeline probably because I am deploying new test models.

```
1: [  FAILED  ] ModelTests/ModelTest.Run/cpu__models_zoo_opset7_ResNet101_DUC_HDC_ResNet101DUC7, where GetParam() = L"cpu_..\\models\\zoo\\opset7\\ResNet101_DUC_HDC\\ResNet101-DUC-7.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/cpu__models_zoo_opset12_ResNet101_DUC_HDC12_ResNet101DUC12, where GetParam() = L"cpu_..\\models\\zoo\\opset12\\ResNet101_DUC_HDC-12\\ResNet101-DUC-12.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/cpu__models_zoo_opset11_FCN_ResNet101_model, where GetParam() = L"cpu_..\\models\\zoo\\opset11\\FCN ResNet-101\\model.onnx"
1: [  FAILED  ] ModelTests/ModelTest.Run/cpu__models_zoo_opset10_SSD_model, where GetParam() = L"cpu_..\\models\\zoo\\opset10\\SSD\\model.onnx"
```
I will revert the deployment and address them.

File: cmake/external/gsl.cmake [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    GIT_TAG a3534567187d2edc428efd3f13466ff75fe5805c  # v4.0.0
+    GIT_SHALLOW ON
+    )
+
+FetchContent_MakeAvailable(GSL)
+
+set(GSL_TARGET "Microsoft.GSL::GSL")
+set(GSL_INCLUDE_DIR "$<TARGET_PROPERTY:${GSL_TARGET},INTERFACE_INCLUDE_DIRECTORIES>")
```
Comment (yuslepukhin) [inline cmake/external/gsl.cmake:16]: >set(GSL_INCLUDE_ [](http://example.com/codeflow?start=0&length=16)

Let's not forget about natvis. Once experienced, it is hard to go back.

Comment (snnn) [reply cmake/external/gsl.cmake:16]: A puppy love!

Yes. The official GSL repo has it in their cmake targets.

Comment (edgchen1) [reply cmake/external/gsl.cmake:16]: Yep, it should be included if we link to the target:
https://github.com/microsoft/GSL/blob/a3534567187d2edc428efd3f13466ff75fe5805c/cmake/guidelineSupportLibrary.cmake#L79-L82

File: cmake/CMakeLists.txt (hunk 4 above)
Comment (edgchen1) [reply cmake/CMakeLists.txt]: we discussed this offline. this particular issue is only applicable to older versions of CMake (<3.8). ORT has a higher minimum version.
so will just stick to using the GSL CMake target for now where applicable.

**Comment by snnn** [comment]
LGTM. Please also let @pranavsharma  take a look.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h [hunk 7]
```diff
@@ -3,10 +3,10 @@
 
 #pragma once
 
+#include "core/common/span_utils.h"
 #include "contrib_ops/cpu/transformers/generation_shared.h"  // for DEBUG_GENERATION
 #include "contrib_ops/cpu/transformers/beam_search_impl_base.h"
 #include "contrib_ops/cpu/transformers/subgraph_t5_encoder.h"
```
Comment (edgchen1) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_t5.h:9]: add back `#include "contrib_ops/cpu/transformers/subgraph_t5_decoder.h"`

File: onnxruntime/test/onnx/tensorprotoutils.cc [hunk 8]
```diff
@@ -6,7 +6,7 @@
 #include <memory>
 #include <algorithm>
 #include <limits>
-#include <gsl/gsl>
+#include "core/common/gsl.h"
```
Comment (edgchen1) [inline onnxruntime/test/onnx/tensorprotoutils.cc:9]: remove this

File: onnxruntime/test/optimizer/initializer_test.cc [hunk 9]
```diff
@@ -80,7 +80,8 @@ TEST(OptimizerInitializerTest, LoadExternalData) {
 
         if (offset + length <= tensor_data_span.size()) {
           Initializer i(tensor_proto, tensor_data_dir_path);
-          EXPECT_EQ(gsl::make_span(i.data<int32_t>(), i.size()), tensor_data_span.subspan(offset, length));
+          const auto* i_data = i.data<int32_t>();
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/initializer_test.cc:83]: could be clearer:
`gsl::span<const int> i_data_span{i.data<int32_t>(), i.size()};`

File: orttraining/orttraining/core/framework/pipeline.h [hunk 10]
```diff
@@ -12,7 +12,7 @@
 #include <thread>
 #include <unordered_map>
 
-#include "gsl/gsl"
+#include "core/common/gsl.h"
```
Comment (edgchen1) [inline orttraining/orttraining/core/framework/pipeline.h:15]: remove this

File: orttraining/orttraining/training_ops/cuda/nn/batch_norm_grad.h [hunk 11]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "gsl/gsl"
+#include "core/common/gsl.h"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/nn/batch_norm_grad.h:6]: remove this

File: orttraining/orttraining/training_ops/cuda/nn/batch_norm_internal.h [hunk 12]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "gsl/gsl"
+#include "core/common/gsl.h"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cuda/nn/batch_norm_internal.h:6]: remove this

File: orttraining/orttraining/training_ops/rocm/nn/batch_norm_grad.h [hunk 13]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "gsl/gsl"
+#include "core/common/gsl.h"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/rocm/nn/batch_norm_grad.h:6]: remove

File: orttraining/orttraining/training_ops/rocm/nn/batch_norm_internal.h [hunk 14]
```diff
@@ -3,7 +3,7 @@
 
 #pragma once
 
-#include "gsl/gsl"
+#include "core/common/gsl.h"
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/rocm/nn/batch_norm_internal.h:6]: remove

**Review by pranavsharma — APPROVED** [review]
LGTM. Hopefully this has no perf impact. Let's keep an eye on the dashboard just in case.
